esutil/bulk_indexer.go

```diff
@@ -0,0 +1,509 @@
+	w.aux = w.aux[:0]
+	w.buf.WriteRune(':')
+	w.buf.WriteRune('{')
+	if item.DocumentID != "" {
+		w.buf.WriteString(`"_id":`)
+		w.aux = strconv.AppendQuote(w.aux, item.DocumentID)
+		w.buf.Write(w.aux)
+		w.aux = w.aux[:0]
+	}
+	w.buf.WriteRune('}')
+	w.buf.WriteRune('}')
+	w.buf.WriteRune('\n')
+	return nil
+}
+
+// writeBody writes the item body to the buffer; it must be called under a lock.
+//
+func (w *worker) writeBody(item BulkIndexerItem) error {
+	// TODO(karmi): Handle errors
+	if item.Body != nil {
+		if _, err := w.buf.ReadFrom(item.Body); err != nil {
+			return err
+		}
+		w.buf.WriteRune('\n')
+	}
+	return nil
+}
+
+// flush writes out the worker buffer; it must be called under a lock.
+//
+func (w *worker) flush() error {
+	if w.isFlushing {
+		if w.bi.config.DebugLogger != nil {
+			w.bi.config.DebugLogger.Printf("[worker-%03d] Flush: Already flushing\n", w.id)
+		}
+		return nil
+	}
+
+	if w.buf.Len() < 1 {
+		if w.bi.config.DebugLogger != nil {
+			w.bi.config.DebugLogger.Printf("[worker-%03d] Flush: Buffer empty\n", w.id)
+		}
+		return nil
+	}
+
+	var (
+		err error
+		blk BulkIndexerResponse
+		ctx = context.Background()
+	)
+
+	defer func() {
+		w.items = w.items[:0]
+		w.buf.Reset()
+		w.isFlushing = false
+	}()
+
+	w.isFlushing = true
+
+	if w.bi.config.DebugLogger != nil {
+		w.bi.config.DebugLogger.Printf("[worker-%03d] Flush: %s\n", w.id, w.buf.String())
+	}
+
+	atomic.AddUint64(&w.bi.stats.numRequests, 1)
+	req := esapi.BulkRequest{
+		Index: w.bi.config.Index,
+		Body:  w.buf,
+
+		Pipeline:            w.bi.config.Pipeline,
+		Refresh:             w.bi.config.Refresh,
+		Routing:             w.bi.config.Routing,
+		Source:              w.bi.config.Source,
+		SourceExcludes:      w.bi.config.SourceExcludes,
+		SourceIncludes:      w.bi.config.SourceIncludes,
+		Timeout:             w.bi.config.Timeout,
+		WaitForActiveShards: w.bi.config.WaitForActiveShards,
+
+		Pretty:     w.bi.config.Pretty,
+		Human:      w.bi.config.Human,
+		ErrorTrace: w.bi.config.ErrorTrace,
+		FilterPath: w.bi.config.FilterPath,
+		Header:     w.bi.config.Header,
+	}
+	res, err := req.Do(ctx, w.bi.config.Client)
+	if err != nil {
+		atomic.AddUint64(&w.bi.stats.numFailed, uint64(len(w.items)))
+		// TODO(karmi): Wrap error
+		return fmt.Errorf("flush: %s", err)
+	}
+	if res.Body != nil {
+		defer res.Body.Close()
+	}
+	if res.IsError() {
+		atomic.AddUint64(&w.bi.stats.numFailed, uint64(len(w.items)))
+		// TODO(karmi): Wrap error (include response struct)
+		return fmt.Errorf("flush: %s", res.String())
+	}
+
+	if err := w.bi.config.Decoder.UnmarshalFromReader(res.Body, &blk); err != nil {
+		// TODO(karmi): Wrap error (include response struct)
+		return fmt.Errorf("flush: error parsing response body: %s", err)
+	}
+
+	for i, blkItem := range blk.Items {
+		var (
+			item BulkIndexerItem
+			info BulkIndexerResponseItem
+			op   string
+		)
+
+		item = w.items[i]
+		for k, v := range blkItem {
+			op = k
+			info = v
+		}
```

Comment:
I am confused by this range loop, shouldn't that contain the code below? The way it is now it loops over the `blkItem`which is a `map[string]BulkIndexerResponseItem`, and sets `op` and `info`; maps don't have a defined order, so it is undefined what `op` and `info` is set to after the loop, and only the last entry will be processed further.
It implicitly works as there is always only one entry in the map per `blkItem`, but this is hard to read and could easily lead to bugs if the response object was created in a different way. If it is intended to always only contain one item and you want to add a label to the `BulkIndexerResponseItem` I suggest to create a dedicated struct holding those two pieces of information rather than creating a map.

Reply:
This is due to the [_bulk](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-bulk.html) result format. For each item, you get back a map with a single key/value pair, where the key is the operation name.

Reply:
Ah, I overlooked that this is due to the ES bulk response, that makes sense then. Resolving this conversation.

Reply:
> It implicitly works as there is always only one entry in the map per blkItem (..)

The Bulk API response is a map with only a single key, and I don't think that can change, since it corresponds to the operation, ie. for a "create" operation, you'll get back a "create" key. So it is kind of "contractually correct" :) I agree the code is a bit too hard-to-read here. I'll think about this some more.

Reply:
I've added a comment to make it more clear for future readers in 0212193.